pulsar-broker/src/main/java/org/apache/pulsar/broker/resourcegroup/ResourceGroupNamespaceConfigListener.java

```diff
@@ -0,0 +1,162 @@
+ */
+package org.apache.pulsar.broker.resourcegroup;
+
+import static org.apache.pulsar.broker.admin.ZkAdminPaths.POLICIES;
+import static org.apache.pulsar.common.policies.path.PolicyPath.path;
+import java.util.function.Consumer;
+import org.apache.pulsar.broker.PulsarService;
+import org.apache.pulsar.broker.resources.NamespaceResources;
+import org.apache.pulsar.broker.resources.TenantResources;
+import org.apache.pulsar.client.admin.PulsarAdminException;
+import org.apache.pulsar.common.policies.data.Policies;
+import org.apache.pulsar.metadata.api.Notification;
+import org.apache.pulsar.metadata.api.NotificationType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ResourceGroupNamespaceConfigListener implements Consumer<Notification> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ResourceGroupNamespaceConfigListener.class);
+    private final ResourceGroupService rgService;
+    private final PulsarService pulsarService;
+    private final NamespaceResources namespaceResources;
+    private final TenantResources tenantResources;
+    private final ResourceGroupConfigListener rgConfigListener;
+
+    public ResourceGroupNamespaceConfigListener(ResourceGroupService rgService, PulsarService pulsarService,
+            ResourceGroupConfigListener rgConfigListener) {
+        this.rgService = rgService;
+        this.pulsarService = pulsarService;
+        this.namespaceResources = pulsarService.getPulsarResources().getNamespaceResources();
+        this.tenantResources = pulsarService.getPulsarResources().getTenantResources();
+        this.rgConfigListener = rgConfigListener;
+        loadAllNamespaceResourceGroups();
+        this.namespaceResources.getStore().registerListener(this);
+    }
+
+    private void updateNamespaceResourceGroup(String path) {
+        String nsName = path.substring(path(POLICIES).length() + 1);
+
+        namespaceResources.getAsync(path).whenCompleteAsync((optionalPolicies, ex) -> {
+            if (ex != null) {
+                LOG.error("Exception when getting namespace {}", nsName, ex);
+                return;
+            }
+            Policies policy = optionalPolicies.get();
+            reconcileNamespaceResourceGroup(nsName, policy);
+        });
+    }
+
+    private void loadAllNamespaceResourceGroups() {
+        tenantResources.getChildrenAsync(path(POLICIES)).whenComplete((tenantList, ex) -> {
+            if (ex != null) {
+                LOG.error("Exception when fetching tenants", ex);
+                return;
+            }
+            for (String ts: tenantList) {
+                namespaceResources.getChildrenAsync(path(POLICIES, ts)).whenComplete((nsList, ex1) -> {
+                    if (ex1 != null) {
+                        LOG.error("Exception when fetching namespaces", ex1);
+                    } else {
+                        for (String ns: nsList) {
+                            updateNamespaceResourceGroup(path(POLICIES, ts, ns));
```

Comment:
What if some resource groups got updated and others did not get updated successfully.  Could there be unintended behavior in that scenario?

Reply:
is the question that in the following scenario:
- namespace NS1 and namespace NS2 are updated to attach to some resource groups.
- broker was able to successfully process NS1, but the NS2 failed (for some reason).

is that correct?
I don't know why NS2 update will fail as all it is doing is updating some internal data structures (unless broker crashes due to some issue). Or there is some bug. If former, upon restart broker will process the config again. If there is a bug, we need to fix it. And to detect any failure in operations, I have added logs indicating the failure and the reason for the failure.

I am also not clear if the scenario is different for other configuration changes handled by broker.

can you please let me know if I misunderstood the question. thank you!

Reply:
@bharanic-dev updateNamespaceResourceGroup does not update anything in zk?